drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1564 @@
+    {
+      return err;
+    }
+
+  priv->mod = mod;
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_getmod
+ *
+ * Description:
+ *    Get the modulation type of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getmod(FAR struct rn2xx3_dev_s *priv, FAR char *mod)
+{
+  ssize_t length;
+  char response[12];
+
+  length =
+      rn2xx3_getparam(priv, "radio get mod\r\n", response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Copy over modulation */
+
+  memcpy(mod, response, length);
+  return 0;
+}
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+
+/****************************************************************************
+ * Name: rn2xx3_open
+ ****************************************************************************/
+
+static int rn2xx3_open(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2xx3_dev_s *priv = inode->i_private;
+  int err;
+
+  err = nxmutex_lock(&priv->devlock);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  /* Only one user at a time */
+
+  if (priv->crefs > 0)
+    {
+      err = -EBUSY;
+      wlerr("Too many fds");
+      goto early_ret;
+    }
+
+  priv->crefs++;
+  wlinfo("crefs: %d\n", priv->crefs);
+
+early_ret:
+  nxmutex_unlock(&priv->devlock);
+  return err;
+}
+
+#endif
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+
+/****************************************************************************
+ * Name: rn2xx3_close
+ ****************************************************************************/
+
+static int rn2xx3_close(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2xx3_dev_s *priv = inode->i_private;
+  int err;
+
+  err = nxmutex_lock(&priv->devlock);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  priv->crefs--;
+  wlinfo("crefs: %d\n", priv->crefs);
+  nxmutex_unlock(&priv->devlock);
+  return err;
+}
+
+#endif
+
+/****************************************************************************
+ * Name: rn2xx3_read
+ *
+ * Description:
+ *   Puts the RN2xx3 into continuous receive mode and waits for data to be
+ *   received before returning it.
+ *
+ ****************************************************************************/
+
+static ssize_t rn2xx3_read(FAR struct file *filep, FAR char *buffer,
+                           size_t buflen)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2xx3_dev_s *priv = inode->i_private;
+  ssize_t length = 0;
+  ssize_t received = 0;
+  static const char receive_cmd[] = "radio rx 0\r\n";
+  char response[30];
+  char hex_byte[3] =
+  {
+    0, 0 /* Extra space for null terminator */
+  };
+
+  /* Exclusive access */
+
+  length = nxmutex_lock(&priv->devlock);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* If we're still marked as receiving and the file position in non-zero, we
```

Comment:
Move these to specific if blocks and last comment after if blocks. Took me way too much brain power to know what's what :)

Reply:
Thank you for your detailed reviews!